### PR TITLE
Tenplay update

### PR DESCRIPTION
**Pull request by king-millez**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

The purpose of this *pull request* is to fix an issue in the `tenplay` extractor, where the script would not correctly obtain an authentication token before attempting to download a video, resulting in a `403` error. I've solved this by requiring a username and password to generate the token, and then providing it when downloading the `m3u8` playlist.

I've also made some minor adjustments to the metadata that is extracted, as the `tenplay` API has been updated and the `json` keys changed. (e.g 'picture' has been changed to `imageUrl` for show thumbnails).

### CODE REVIEW

File: yt_dlp/extractor/tenplay.py [hunk 1]
```diff
@@ -1,70 +1,76 @@
...
-            # 'ie_key': 'BrightcoveNew',
+            'id': content_id,
+            'title': data.get('title'),
+            'description': data.get('description'),
+            'age_limit': parse_age_limit(data.get('classification')),
+            'series': data.get('showName'),
+            'season': data.get('showContentSeason'),
+            'timestamp': data.get('published'),
+            'thumbnail': data.get('imageUrl'),
+            'uploader_id': 'Channel 10',
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:75]: please explain

File: yt_dlp/extractor/tenplay.py [hunk 2]
```diff
@@ -1,70 +1,76 @@
...
         data = self._download_json(
-            'https://10play.com.au/api/video/' + content_id, content_id)
-        video = data.get('video') or {}
-        metadata = data.get('metaData') or {}
-        brightcove_id = video.get('videoId') or metadata['showContentVideoId']
-        # brightcove_url = smuggle_url(
-        #     self.BRIGHTCOVE_URL_TEMPLATE % brightcove_id,
-        #     {'geo_countries': ['AU']})
+            'https://10play.com.au/api/v1/videos/' + content_id, content_id)
+        _video_url = self._download_json(data.get('playbackApiEndpoint'), content_id, 'Downloading video JSON', headers={'Authorization': _token}).get('source')
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:57]: split long line

```suggestion
        _video_url = self._download_json(
            data.get('playbackApiEndpoint'), content_id, 'Downloading video JSON',
            headers={'Authorization': _token}).get('source')
```

File: yt_dlp/extractor/tenplay.py [hunk 3]
```diff
@@ -1,70 +1,76 @@
...
     _GEO_BYPASS = False
-    _FASTLY_URL_TEMPL = 'https://10-selector.global.ssl.fastly.net/s/kYEXFC/media/%s?mbr=true&manifest=m3u&format=redirect'
+
+    def _get_bearer_token(self, video_id):
+        _authdata = self._get_login_info()
+        _time = datetime.now()
+        _auth_header = base64.b64encode(f'{_time.year}{"{:02d}".format(_time.month)}{"{:02d}".format(_time.day)}000000'.encode('ascii')).decode('ascii')
+        if(_authdata[0] is None or _authdata[1] is None):
+            raise Exception('Your 10play account\'s details must be provided with --username and --password.')
+        username, password = self._get_login_info()
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:43]:
```suggestion
        username, password = self._get_login_info()
        _time = datetime.now()
        _auth_header = base64.b64encode(f'{_time.year}{"{:02d}".format(_time.month)}{"{:02d}".format(_time.day)}000000'.encode('ascii')).decode('ascii')
        if username is None or password is None:
            self.raise_login_required('Your 10play account\'s details must be provided with --username and --password.')
```

**Review by pukkandan — COMMENTED** [review]
One more thing:
Py2 support has currently only been deprecated and not removed. So you either need to make the code python2 compatible, or I will postpone merging this PR untill after the support is removed (in a couple of releases)

File: yt_dlp/extractor/tenplay.py [hunk 4]
```diff
@@ -1,70 +1,76 @@
...
-            'timestamp': 1240828200,
-            'upload_date': '20090427',
-            'uploader_id': '2199827728001',
+            'title': "Todd Sampson's Body Hack - S4 Ep. 2",
+            'description': 'md5:a73ea55671034c6367784405423ef5a0',
+            'age_limit': 15,
+            'timestamp': 1601379660,
+            'upload_date': '20200909',
+            'uploader_id': 'Channel 10',
         },
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:27]: Don't change test without good reason. Add as a new test case if needed. Also, the `id` of videos must be same as before so as not to break `--download-archive`

File: yt_dlp/extractor/tenplay.py [hunk 5]
```diff
@@ -1,70 +1,76 @@
 # coding: utf-8
 from __future__ import unicode_literals
-
+from datetime import datetime
+import base64
 from .common import InfoExtractor
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:5]: need empty line

```suggestion

from .common import InfoExtractor
```

File: yt_dlp/extractor/tenplay.py [hunk 6]
```diff
@@ -1,70 +1,76 @@
 # coding: utf-8
 from __future__ import unicode_literals
-
+from datetime import datetime
+import base64
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:4]:
```suggestion

from datetime import datetime
import base64

```

seperate into groups

File: yt_dlp/extractor/tenplay.py (hunk 1 above)
Comment (king-millez) [reply yt_dlp/extractor/tenplay.py:75]: I can put `Channel 10` into the `uploader` entry, and dynamically change `uploader_id` to a value extracted from the `m3u8`, as it was incorrectly hard-coded before.

File: yt_dlp/extractor/tenplay.py [hunk 7]
```diff
@@ -35,26 +37,28 @@ class TenPlayIE(InfoExtractor):
...
-        _time = datetime.now()
-        _auth_header = base64.b64encode(f'{_time.year}{"{:02d}".format(_time.month)}{"{:02d}".format(_time.day)}000000'.encode('ascii')).decode('ascii')
-        if(_authdata[0] is None or _authdata[1] is None):
-            raise Exception('Your 10play account\'s details must be provided with --username and --password.')
         username, password = self._get_login_info()
+        _time = datetime.now()
+        _timestamp = str(_time.year) + str("{:02d}".format(_time.month)) + str("{:02d}".format(_time.day)) + '000000'
+        _auth_header = base64.b64encode(_timestamp.encode('ascii')).decode('ascii')
+        if username is None or password is None:
+            self.raise_login_required('Your 10play account\'s details must be provided with --username and --password.')
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:45]: can be simplified to:
```suggestion
        if username is None or password is None:
            self.raise_login_required('Your 10play account\'s details must be provided with --username and --password.')
        _timestamp = datetime.now().strftime('%Y%M%d000000')
        _auth_header = base64.b64encode(_timestamp.encode('ascii')).decode('ascii')
```

File: yt_dlp/extractor/tenplay.py [hunk 8]
```diff
@@ -1,70 +1,81 @@
...
+            'id': content_id,
+            'title': data.get('title'),
+            'description': data.get('description'),
+            'age_limit': parse_age_limit(data.get('classification')),
+            'series': data.get('showName'),
+            'season': data.get('showContentSeason'),
+            'timestamp': data.get('published'),
+            'thumbnail': data.get('imageUrl'),
+            'uploader': 'Channel 10',
+            'uploaer_id': '2199827728001',
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:80]:
```suggestion
            'uploader_id': '2199827728001',
```
typo

File: yt_dlp/extractor/tenplay.py (hunk 4 above)
Comment (pukkandan) [reply yt_dlp/extractor/tenplay.py]: looks like they no longer use brightcove. So it cant be helped

**Comment by pukkandan** [comment]
Should fix: https://github.com/ytdl-org/youtube-dl/issues/28657 https://github.com/ytdl-org/youtube-dl/issues/28641 https://github.com/ytdl-org/youtube-dl/issues/28822 https://github.com/ytdl-org/youtube-dl/issues/29005

Related: https://github.com/ytdl-org/youtube-dl/pull/28728

File: yt_dlp/extractor/tenplay.py [hunk 9]
```diff
@@ -75,7 +77,7 @@ def _real_extract(self, url):
 
         return {
             'formats': formats,
-            'id': content_id,
+            'id': data.get('altId'),
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:80]:
```suggestion
            'id': data.get('altId') or video_id,
```

File: yt_dlp/extractor/tenplay.py [hunk 10]
```diff
@@ -48,7 +49,8 @@ def _get_bearer_token(self, video_id):
         username, password = self._get_login_info()
         if username is None or password is None:
             self.raise_login_required('Your 10play account\'s details must be provided with --username and --password.')
-        _timestamp = datetime.now().strftime('%Y%M%d000000')
+        _timestamp = datetime.now().strftime('%Y%m%d000000')
+        print(_timestamp)
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:53]:
```suggestion
```

File: yt_dlp/extractor/tenplay.py [hunk 11]
```diff
@@ -1,70 +1,89 @@
...
             'formats': formats,
-            'id': brightcove_id,
-            'title': video.get('title') or metadata.get('pageContentName') or metadata['showContentName'],
-            'description': video.get('description'),
-            'age_limit': parse_age_limit(video.get('showRatingClassification') or metadata.get('showProgramClassification')),
-            'series': metadata.get('showName'),
-            'season': metadata.get('showContentSeason'),
-            'timestamp': parse_iso8601(metadata.get('contentPublishDate') or metadata.get('pageContentPublishDate')),
-            'thumbnail': video.get('poster'),
+            'id': data.get('altId') or video_id,
```
Comment (pukkandan) [inline yt_dlp/extractor/tenplay.py:79]:
```suggestion
            'id': data.get('altId') or content_id,
```

**Pull request merged**